src/modules/cmdpal/Microsoft.CmdPal.UI.ViewModels/ContextMenuViewModel.cs

```diff
@@ -0,0 +1,267 @@
+    /// include the secondary one). This map can be used to quickly check if a
+    /// shortcut key was pressed
+    /// </summary>
+    /// <returns>a dictionary of KeyChord -> Context commands, for all commands
+    /// that have a shortcut key set.</returns>
+    public Dictionary<KeyChord, CommandContextItemViewModel> Keybindings()
+    {
+        if (CurrentContextMenu == null)
+        {
+            return [];
+        }
+
+        return CurrentContextMenu
+            .OfType<CommandContextItemViewModel>()
+            .Where(c => c.HasRequestedShortcut)
+            .ToDictionary(
+            c => c.RequestedShortcut ?? new KeyChord(0, 0, 0),
+            c => c);
+    }
+
+    public ContextKeybindingResult? CheckKeybinding(bool ctrl, bool alt, bool shift, bool win, VirtualKey key)
+    {
+        var keybindings = Keybindings();
+        if (keybindings != null)
+        {
+            // Does the pressed key match any of the keybindings?
+            var pressedKeyChord = KeyChordHelpers.FromModifiers(ctrl, alt, shift, win, key, 0);
+            if (keybindings.TryGetValue(pressedKeyChord, out var item))
+            {
+                return InvokeCommand(item);
+            }
+        }
+
+        return null;
+    }
+
+    public bool CanPopContextStack()
+    {
+        return ContextMenuStack.Count > 1;
+    }
+
+    public void PopContextStack()
+    {
+        if (ContextMenuStack.Count > 1)
+        {
+            ContextMenuStack.RemoveAt(ContextMenuStack.Count - 1);
+        }
+
+        OnPropertyChanging(nameof(CurrentContextMenu));
+        OnPropertyChanged(nameof(CurrentContextMenu));
+
+        ListHelpers.InPlaceUpdateList(FilteredItems, [.. CurrentContextMenu!]);
+    }
+
+    private void PushContextStack(IEnumerable<IContextItemViewModel> commands)
+    {
+        ContextMenuStack.Add(commands.ToList());
+        OnPropertyChanging(nameof(CurrentContextMenu));
+        OnPropertyChanged(nameof(CurrentContextMenu));
+
+        ListHelpers.InPlaceUpdateList(FilteredItems, [.. CurrentContextMenu!]);
+    }
+
+    private void ResetContextMenu()
+    {
+        while (ContextMenuStack.Count > 1)
+        {
+            ContextMenuStack.RemoveAt(ContextMenuStack.Count - 1);
+        }
+
+        OnPropertyChanging(nameof(CurrentContextMenu));
+        OnPropertyChanged(nameof(CurrentContextMenu));
+
+        ListHelpers.InPlaceUpdateList(FilteredItems, [.. CurrentContextMenu!]);
+    }
+
+    public ContextKeybindingResult InvokeCommand(CommandItemViewModel? command)
+    {
+        if (command == null)
+        {
+            return ContextKeybindingResult.Unhandled;
+        }
+
+        if (command.HasMoreCommands)
+        {
+            // Execute the command
+            WeakReferenceMessenger.Default.Send<PerformCommandMessage>(new(command.Command.Model, command.Model));
+
+            // Display the commands child commands
+            PushContextStack(command.MoreCommands);
```

Comment:
⛔ This might need to be `.AllCommands` I think. The nested menu on "We can go deeper..." should also include the "Do 3" command that's actually on that item